Fix compatibility with Emscripten 3.1.28

kleisauke opened the pull request:
Also, take advantage of `EM_JS_DEPS` while we are here (commit f0e8a0efb5418c345e6fefa882e5fad5d1e0e401).

Tested with:
```console
$ sed -i'.old' "/^NODE_JS = /s/emsdk.*/\[&, '--experimental-wasm-bigint'\]/" $EMSDK/.emscripten
$ autoreconf -fiv
$ emconfigure ./configure --host=wasm32-unknown-linux --enable-static --disable-shared --disable-builddir --disable-multi-os-directory --disable-raw-api --disable-docs CFLAGS="-DWASM_BIGINT"
$ make
$ EMMAKEN_JUST_CONFIGURE=1 emmake make check RUNTESTFLAGS="LDFLAGS_FOR_TARGET='-sEXPORTED_FUNCTIONS=_main,_malloc,_free -sALLOW_TABLE_GROWTH -sWASM_BIGINT'"
...
		=== libffi Summary ===

# of expected passes		1388
# of unexpected failures	70
# of unexpected successes	2
# of expected failures		2
# of unresolved testcases	2
# of unsupported tests		28
```

Context: https://github.com/emscripten-core/emscripten/pull/18391#issuecomment-1364517399.

/cc @sbc100

kleisauke committed:
Fix compatibility with Emscripten 3.1.28

kleisauke committed:
Take advantage of `EM_JS_DEPS`

Available since Emscripten 3.1.24.

sbc100 commented on build.sh:
```diff
@@ -38,6 +38,8 @@ export LDFLAGS="-L$TARGET/lib -O3 -sEXPORTED_FUNCTIONS=_main,_malloc,_free -sALL
   export LDFLAGS+=" -sWASM_BIGINT"
 fi
+# Ensure the `ffi_call` symbol can be found at link time
+export LDFLAGS+=" -sLLD_REPORT_UNDEFINED=0"
```
This is an unfortunate thing to have to add.  

Why is this symbol defined?  Is it and EM_JS symbol?  Do you know where the undefined reference coming from?

kleisauke replied:
Good question. It is defined here:
https://github.com/hoodmane/libffi-emscripten/blob/8bcb4262bfc8cb57b6314b93143f18c650f63492/src/wasm32/ffi.c#L173
(i.e., as an `EM_JS` symbol)

> Do you know where the undefined reference coming from?

I'm not sure, let me investigate that. Perhaps defining an `ffi_call` stub in native code fixes this.

kleisauke replied:
... relevant header declaration:
https://github.com/hoodmane/libffi-emscripten/blob/8bcb4262bfc8cb57b6314b93143f18c650f63492/include/ffi.h.in#L482-L486

sbc100 replied:
I think I see what is happening.  EM_JS functions are functions that are defined in native code.  We they kind of need to be undefined when wasm-ld runs, so it knows to generate imports for these symbols.

When an EM_JS functions is used within the same file where it is declared this is not an issues since the macro declares the symbol with an `import_name` attribute, which tells the linker its OK to be undefined/imported.

As you suggested I think the simplest solution would be write a native wrapper for the EM_JS function in the same file where it is defined.

kleisauke replied:
Native wrapper function implemented with commit d1798d88fb955900b2452af3c28f487fa85b6808, seems to work!

kleisauke committed:
nit: ensure correct placement of pointer asterisks

kleisauke committed:
CI: remove use of `EM_CONFIG` env

See commit:
https://github.com/emscripten-core/emsdk/commit/3d87d5ea8143b3636f872fb05b896eb4a19a070b

kleisauke commented:
Marking as draft due to dependence on PR https://github.com/emscripten-core/emscripten/pull/18438 (which would avoid the need of commit 42fdeeb25e6993ce5b146727c83d962434a7fab3).

After that, this PR can be re-titled as "Fix compatibility with Emscripten 3.1.28".

kleisauke commented:
Just dropped commit 42fdeeb25e6993ce5b146727c83d962434a7fab3 and 8bcb4262bfc8cb57b6314b93143f18c650f63492. This is ready for review now.

Note that commit f0e8a0efb5418c345e6fefa882e5fad5d1e0e401 would mean that Emscripten 3.1.24 is the minimum required version to compile this. Let me know if you want a separate PR for that.

sbc100 commented on src/wasm32/ffi.c:
```diff
@@ -170,7 +170,7 @@ unbox_small_structs, (ffi_type type_ptr), {
 EM_JS_MACROS(
 void,
-ffi_call, (ffi_cif * cif, ffi_fp fn, void *rvalue, void **avalue),
+ffi_call_helper, (ffi_cif *cif, ffi_fp fn, void *rvalue, void **avalue),
```
Maybe call this `ffi_call_js`?

kleisauke replied:
Good point, perhaps that's clearer. I aligned/borrowed this naming convention with those other `EM_JS` functions.

@hoodmane WDYT? Should we use `*_js` here (and possibly those other occurrences) instead of `*_helper`?

hoodmane replied:
+1 for `ffi_call_js`.

hoodmane replied:
I cherry picked this commit using `ffi_call_js`.

kleisauke replied:
I just opened PR https://github.com/hoodmane/libffi-emscripten/pull/22 for this.

sbc100 reviewed (commented):
emscripten stuff lgtm

kleisauke commented:
> Note that commit [f0e8a0e](https://github.com/hoodmane/libffi-emscripten/commit/f0e8a0efb5418c345e6fefa882e5fad5d1e0e401) would mean that Emscripten 3.1.24 is the minimum required version to compile this.

Thinking about this further, I think we could make the `EM_JS_DEPS` macro no-op with something like this:
```c
/* The __EMSCRIPTEN_major__ etc macros used to be passed implicitly, version
 * 3.1.4 moved them to a version header and version 3.1.23 dropped the
 * backwards compatibility. To work consistently on all versions, include the
 * header only if the version macros aren't present.
 * https://github.com/emscripten-core/emscripten/commit/f99af02045357d3d8b12e63793cef36dfde4530a
 * https://github.com/emscripten-core/emscripten/commit/f76ddc702e4956aeedb658c49790cc352f892e4c
 */
#ifndef __EMSCRIPTEN_major__
#include <emscripten/version.h>
#endif

/* Make the EM_JS_DEPS macro no-op before Emscripten 3.1.24.
 */
#if __EMSCRIPTEN_major__*10000 + __EMSCRIPTEN_minor__*100 + __EMSCRIPTEN_tiny__ < 30124
#define EM_JS_DEPS(tag, deps)
#endif
```
(untested)

hoodmane committed:
Merge branch 'libffi-emscripten-main' into tot-compat

hoodmane commented:
I'm going to close this. I merged some of it and other stuff is superseded now. Thanks @kleisauke!

kleisauke commented:
Thanks! BTW, do you think we should enforce `-sWASM_BIGINT`? It might make the code a bit easier and avoids having to specify  `-DWASM_BIGINT` during compilation.
https://github.com/hoodmane/libffi-emscripten/blob/e2f8291ab03c351347a0b4d1e992e125b42a31a4/testsuite/emscripten/node-tests.sh#L25-L26

The projects depending on this repo (i.e. Pyodide/Python ctypes,  wasm-vips, etc.) already builds with `-sWASM_BIGINT` by default (afaik).

hoodmane commented:
Yeah it's a lot of work to support the no wasm bigint case. @pmp-p seems to be an evangelist for not using wasm-bigint but I'm not exactly sure why. compilation, runtime, and size are all notably less with WASM_BIGINT integration and I think support is missing in <3% of browsers in the wild. I know the godot game engine builds without it (but it doesn't use libffi).

For a long time we only thought we were supporting no wasm bigint since we were testing against browsers with js bigint integration. But the no wasm bigint case adds a lot of really complicated code:
https://github.com/hoodmane/libffi-emscripten/commit/e7582c7133c10aebca011a218beb91c653d7f063

hoodmane commented:
By the way @kleisauke I would appreciate a PR adding the support for Emscripten < 3.1.24.

pmp-p commented:
> @pmp-p seems to be an evangelist for not using wasm-bigint

also add to that shared memory,  atomics and threads, simd/neon ... 

The reason is very simple, there are litterally truckload of devices out there stuck at different level of wasm implementation because they won't receive browser updates anymore. But when wasm is present and not behing a flag, it is obviously MVP 1.0 : so instead of wondering who has implemented feature or not just use the established standard.

Sticking to MVP is a way to stop stockpiling perfectly useable devices ( wasm is quite safe on unsafe devices ) in landfills


Another reason  is for evaluating Wasi runtimes in a fair manner. As they don't all implement extra outside the mvp especially when it comes to embedding ( more features == more strain on device )

that's also why i proposed to have two type of wasm wheels on pypi,  mvp (cpython-wasm) and bleeding edge (pyodide)

hoodmane commented:
I don't think Emscripten supports targeting wasm mvp 1.0 at all, it always requires the mutable globals extension.

I am in favor of keeping support for no-wasm-bigint in this libffi port. This way if no-wasm-bigint Python builds want ctypes, they can have it. 

There is a tradeoff between compatibility on the one hand and on the other hand performance and maintainability. Using wasm-bigint significantly reduces our exposure to weird bugs. It's a major issue for Pyodide: frequently when we update compiler versions, some collection of packages start failing to compile. Since we operate on a limited amount of donated effort, it doesn't make sense to expend the effort to avoid requiring a feature that is present in 98% of browsers in the wild.

pmp-p commented:
so you think mvp+bigint+~~mutable globals~~(=> no threads see next post)  should be the absolute minimum requirement and is a safe bet ?

do you remember where you got the info about <3% of browsers that don't support bigint ?

if pyodide can start on Chrome 81.0.4044 i'll change my mind, i'll try today

edit/ it does not,  that would wipe out android kitkat devices, plenty of them still in storage waiting to be sold under $30 piece.

it works on Chrome 95.0.4638 so android 5.1 is saved for now.

So i guess i'll change my mind when a newer Wasm standard is published and kitkat stock has vanished.

sbc100 commented:
> I don't think Emscripten supports targeting wasm mvp 1.0 at all, it always requires the mutable globals extension.

IIUC, you can still target MVP wasm with emscripten, in fact that was the default until recently.  The default was recently changed, but only to match that the new llvm defaults: https://reviews.llvm.org/D125728.  But non-MVP features are not used then when targeting old runtimes: https://github.com/emscripten-core/emscripten/pull/17689

Regarding mutable global specifically, those are only (by emscripten) used in multi-threaded builds.

.  emscripten, by default, will still target MVP.  We do use mutable globals if you 
> I am in favor of keeping support for no-wasm-bigint in this libffi port. This way if no-wasm-bigint Python builds want ctypes, they can have it.
> 
> There is a tradeoff between compatibility on the one hand and on the other hand performance and maintainability. Using wasm-bigint significantly reduces our exposure to weird bugs. It's a major issue for Pyodide: frequently when we update compiler versions, some collection of packages start failing to compile. Since we operate on a limited amount of donated effort, it doesn't make sense to expend the effort to avoid requiring a feature that is present in 98% of browsers in the wild.

hoodmane commented:
Minimum chrome version for wasm bigint is 85 according to https://webassembly.org/roadmap/

Chrome 85 came out in August 2020 so it is only two and a half years old.

My 3% number came from the minimum supported versions on the webassembly roadmap plus can use browser usage statistics. I have no clue what methodology caniuse uses to collect those statistics.
